src/edu/csus/ecs/pc2/ui/ShadowControlPane.java

```diff
@@ -736,6 +876,58 @@ public void run() {
         return testConnectionButton;
     }
 
+    /**
+     * Returns a Map containing the key/value elements in the specified JSON string.
+     * This method uses the Jackson {@link ObjectMapper} to perform the conversion from the JSON
+     * string to a Map.  Note that the ObjectMapper recurses for nested JSON elements, returning
+     * a appropriate Object in the Map under the corresponding key string.
+     * 
+     * @param jsonString a JSON string to be converted to a Map
+     * @return a Map mapping the keys in the JSON string to corresponding values, or null if the input
+     *          String is null or if an exception occurs while converting the JSON to a Map.
+     */
+    @SuppressWarnings("unchecked")
+    protected static Map<String, Object> getMap(String jsonString) {
+        
+        if (jsonString == null){
+            return null;
+        }
+        
+        ObjectMapper mapper = new ObjectMapper();
+        try {
+            Map<String, Object> map = mapper.readValue(jsonString, Map.class);
+            return map;
+        } catch (IOException e) {
+            e.printStackTrace(System.err);
+            return null;
+        }
+    }
+
+    private void addConnectTableEntry(String msg)
+    {
+        SwingUtilities.invokeLater(new Runnable() {
+            public void run() {
+                Object[] objects = new Object[2];
+                
+                try {
+                    GregorianCalendar cal = new GregorianCalendar();
+                    
+                    lastDateFormat.setCalendar(cal);
+                    objects[0] = lastDateFormat.format(cal.getTime());
+                } catch(Exception e) {
+                    objects[0] = "Unknown";
+                }
+                if(msg == null || msg.isEmpty()) {
+                    objects[1] = "<Empty Message>";
+                } else {
+                    objects[1] = msg;
+                }
+                connectStatusTableModel.addRow(objects);
```

Comment:
Rather than add the row, the row should be inserted from the top.  
Unless, as entries are added, the scrolling reveals the last connection logged.

Reply:
I agree with Doug; IMHO each entry should either be added to the TOP of the list (so, it's a Stack), or else adding an entry should force the scrollpane to focus on the bottom (newest) entry.

Reply:
It was a tiny bit more complicated than either of those two solutions.  I opted for the "adjustmentlistener", but that didn't work exactly right either (you couldn't scroll back manually) - so I fixed that, and it now seems to always add to the end, and keep the end visible.  The problem with simply scrolling the scroll bar to the max, was that it would always miss the most recent row, and not show the LAST entry until the next row came in (people had various solutions to this, but the listener seems to most reasonable).